cudapoa/src/main.cpp

```diff
@@ -0,0 +1,277 @@
+    std::vector<BatchSize> list_of_batch_sizes;
+    std::vector<std::vector<int32_t>> list_of_groups_per_batch;
+
+    get_multi_batch_sizes(list_of_batch_sizes, list_of_groups_per_batch, poa_groups, parameters.banded, parameters.consensus_mode == 1);
+
+    int32_t group_count_offset = 0;
+
+    for (int32_t b = 0; b < get_size(list_of_batch_sizes); b++)
+    {
+        auto& batch_size      = list_of_batch_sizes[b];
+        auto& batch_group_ids = list_of_groups_per_batch[b];
+
+        // Initialize batch.
+        std::unique_ptr<Batch> batch = initialize_batch(parameters.consensus_mode == 1, parameters.banded, parameters.gpu_mem_allocation, batch_size);
+
+        // Loop over all the POA groups for the current batch, add them to the batch and process them.
+        int32_t group_count = 0;
+
+        for (int32_t i = 0; i < get_size(batch_group_ids);)
+        {
+            Group& group = poa_groups[batch_group_ids[i]];
+            std::vector<StatusType> seq_status;
+            StatusType status = batch->add_poa_group(seq_status, group);
+
+            // NOTE: If number of batch groups smaller than batch capacity, then run POA generation
+            // once last POA group is added to batch.
+            if (status == StatusType::exceeded_maximum_poas || (i == get_size(batch_group_ids) - 1))
+            {
+                // at least one POA should have been added before processing the batch
+                if (batch->get_total_poas() > 0)
+                {
+                    // No more POA groups can be added to batch. Now process batch.
+                    process_batch(batch.get(), parameters.consensus_mode == 1, true);
+
+                    if (graph_output.good())
+                    {
+                        std::vector<DirectedGraph> graph;
+                        std::vector<StatusType> graph_status;
+                        batch->get_graphs(graph, graph_status);
+                        for (auto& g : graph)
+                        {
+                            graph_output << g.serialize_to_dot() << std::endl;
+                        }
+                    }
+
+                    // After MSA/consensus is generated for batch, reset batch to make room for next set of POA groups.
+                    batch->reset();
+
+                    // In case that number of batch groups is more than the capacity available on GPU, the for loop breaks into smaller number of groups.
+                    // if adding group i in batch->add_poa_group is not successful, it wont be processed in this iteration, therefore we print i-1
+                    // to account for the fact that group i was excluded at this round.
+                    if (status == StatusType::success)
+                    {
+                        std::cout << "Processed groups " << group_count + group_count_offset << " - " << i + group_count_offset << " (batch " << b << ")" << std::endl;
```

Comment:
we probably have to keep these log-type output as `cerr` not to be mixed with printing output data.

Reply:
Done.